spec.bs

```diff
@@ -192,31 +327,59 @@ Specific examples:
-
-<h5 dfn>Change frozenness of document</h5>
-To change the [=Document/frozenness=] of a document given a document |doc| and [=Document/frozenness=] state |x|:
-1. If |x| is <code>true</code>:
-    1. Set |doc|'s [=Document/frozenness=] state to <code>true</code>.
+The UA will typically execute [=change the frozenness of a top-level document=] with false when the user revisits that browsing context. In addition, the UA may choose to periodically execute [=change the frozenness of a top-level document=] with false in the background, if plentiful resources are available.
+
+<h4 id="changing-frozenness">Changing the frozenness of documents</h4>
+
+<div algorithm>
+    To <dfn>change the frozenness of a top-level document</dfn>, given a {{Document}} |topLevelDoc| and boolean [=Document/frozenness=] state |frozenness|:
+
+    1. Assert: |doc|'s [=Document/browsing context=] is a [=top-level browsing context=].
+    1. [=Change the frozenness of a document=] given |topLevelDoc|, |frozenness|, and false.
+    1. Let |descendants| be the [=list of the descendant browsing contexts=] of |doc|.
+    1. For each [=/browsing context=] |b| in |descendants|:
+        1. Let |descendantDocument| be the [=active document=] of |b|.
+        1. [=Change the frozenness of a document=] given |descendantDocument|, |frozenness|, and false.
+</div>
+
+<div algorithm>
+    To <dfn>change the frozenness of a document</dfn>, given a {{Document}} |doc|, a boolean [=Document/frozenness=] state |frozenness|, and a boolean |auto resume frozen media|:
+
+    1. If |frozenness| is true, run the [=freeze steps=] for |doc| given |auto resume frozen media|.
+    1. Otherwise, run the [=resume steps=] given |doc|.
+</div>
+
+<div algorithm>
+    To run the <dfn>freeze steps</dfn> for a {{Document}} <var>doc</var>, given a boolean |auto resume frozen media|:
+
+    1. Set |doc|'s [=Document/frozenness=] state to true.
     1. [=Fire an event=] named <code>freeze</code> at |doc|.
-1. Otherwise:
+    1. Let |elements| be all [=media elements=] in |doc|, in [=tree order=].
+    1. For each |element| in |elements|:
+        1. If |element|'s {{HTMLMediaElement/paused}} is false, then:
+            1. Set |element|'s [=HTMLMediaElement/resume frozen flag=] to |auto resume frozen media|.
+            1. Call |element|'s {{HTMLMediaElement/pause()}} method.
+
+        NOTE: it is intentional that the ordering between the assignment of the of frozneness state
+        occurs first before event firing.
+</div>
+
+<div algorithm>
+    To run the <dfn export>resume steps</dfn> for a {{Document}} <var>doc</var>:
+
+    1. Let |elements| be all [=media elements=] in |doc|, in [=tree order=].
+        1. For each |element| in |elements|:
+            1. If |elements|'s [=HTMLMediaElement/resume frozen flag=] is true.
+                1. Set |elements|'s [=HTMLMediaElement/resume frozen flag=] to false.
+                1. Call |element|'s {{HTMLMediaElement/play()}} method.
```

Comment:
Again you need to not call the public method. It appears this will require more refactoring work on the HTML side. For example, I think all the stuff about the promise being created and returned is not the behavior you want. You'll need to factor out some "internal play steps" in the HTML spec.

Reply:
Done